packages/dynamite/dynamite_runtime/lib/src/http_client.dart

```diff
@@ -2,48 +2,174 @@ import 'dart:async';
+    future = completer.future;
+
+    // ignore: discarded_futures
+    response.then(
+      (final response) async {
+        final rawBody = switch (bodyType) {
+          const FullType(Uint8List) => await response.bytes,
+          const FullType(String) => await response.stringBody,
+          _ => await response.jsonBody,
+        };
+
+        final body = deserializeBody<T>(rawBody, serializers, bodyType);
+        final headers = deserializeHeaders<U>(response.responseHeaders, serializers, headersType);
+
+        _response = DynamiteResponse<T, U>(
+          response.statusCode,
+          body,
+          headers,
+        );
 
-extension DynamiteHttpClientResponseBody on HttpClientResponse {
-  Future<Uint8List> get bodyBytes async {
-    final buffer = BytesBuilder();
+        completer.complete(_response);
+      },
+      onError: completer.completeError,
+    );
+  }
 
-    await forEach(buffer.add);
+  factory DynamiteRawResponse.fromJson(
+    final Map<String, Object?> json, {
+    required final Serializers serializers,
+    final FullType? bodyType,
+    final FullType? headersType,
+  }) {
+    final statusCode = json['statusCode']! as int;
+    final body = deserializeBody<T>(json['body'], serializers, bodyType);
+    final headers = deserializeHeaders<U>(json['headers'], serializers, headersType);
+
+    final response = DynamiteResponse<T, U>(
+      statusCode,
+      body,
+      headers,
+    );
 
-    return buffer.toBytes();
+    return DynamiteRawResponse._fromJson(
+      response,
+      bodyType: bodyType,
+      headersType: headersType,
+      serializers: serializers,
+    );
   }
 
-  Future<String> get body => transform(utf8.decoder).join();
+  DynamiteRawResponse._fromJson(
+    this._response, {
+    required this.bodyType,
+    required this.headersType,
+    required this.serializers,
+  });
+
+  final Serializers serializers;
+
+  final FullType? bodyType;
+
+  final FullType? headersType;
 
-  Future<dynamic> get jsonBody => transform(utf8.decoder).transform(json.decoder).first;
+  late final Future<DynamiteResponse<T, U>> future;
 
-  Map<String, String> get responseHeaders {
-    final responseHeaders = <String, String>{};
-    headers.forEach((final name, final values) {
-      responseHeaders[name] = values.last;
-    });
+  DynamiteResponse<T, U>? _response;
+
+  DynamiteResponse<T, U> get response {
+    final response = _response;
+
+    if (response == null) {
+      throw StateError('The response did not finish yet. Make sure to await `this.future`.');
+    }
 
-    return responseHeaders;
+    return response;
   }
-}
 
-class DynamiteResponse<T, U> {
-  DynamiteResponse(
-    this.data,
-    this.headers,
-  );
+  static T? deserializeBody<T>(final Object? serialized, final Serializers serializers, final FullType? bodyType) {
+    // The serialized value can be correct if we use the more efficient helpers from BytesStreamExtension.
```

Comment:
Our ByteStreamExtension (formerly HttpClientExtension) has three helper methods to get the body data (bodyString, bodyBytes and bodyJson). Especially the first is more effitient than parsing it to json and then to a string.

This condition detects whether the type is already correct (because we used the efficient approach from above) and just returns it.

This also eliminates your concern from above with the serializer being less efficient than a cast as we use the bodyString anyways.

Does this make sense? I'll change the comments if so.

Reply:
I see, please add this information to the comment (and the one below as well).